README.md
# user_auth

Dart - Backend user auth helpe, with common methods

[![Pub](https://img.shields.io/badge/Get%20library-pub-blue)](https://pub.dev/packages/user_auth)
[![Example](https://img.shields.io/badge/Example-Ex-success)](https://pub.dev/packages/user_auth#-example-tab-)

### Show some :heart: and star the repo

[![Fork](https://img.shields.io/github/forks/msayed-net/user_auth?style=social)](https://github.com/msayed-net/user_auth/fork)
[![Star](https://img.shields.io/github/stars/msayed-net/user_auth?style=social)](https://github.com/msayed-net/user_auth/stargazers)
[![Watch](https://img.shields.io/github/watchers/msayed-net/user_auth?style=social)](https://github.com/msayed-net/user_auth/) 

## usage
* import `package:user_auth/user_auth.dart`.
* create instance from `UserAuth`.
* init with `apiBaseUrl` and other parameters.

## methods
* `login` : waits for `user` in response to return. 
* `check` : waits for `user` in response to return.  
* `logout` : return `true` | `false`.
* `loadUser` : return `user` if stored in prefs.
* `register` : waits for `user` in response to return. 

## example
```dart
import 'package:user_auth/user_auth.dart';

// instance
UserAuth user = new UserAuth();

// init
SharedPreferences prefs = await SharedPreferences.getInstance();
  await user.init(
    apiBaseUrl: 'https://example.com/api',
    store: true,
    prefs: prefs, // need if store: true
    loginUrl: "/user/login",
    registerUrl: "/user/register",
    checkUrl: "/user/details",
    logoutUrl: "/user/logout",
  );

// Note : change 'Bearer ' and api_token with your own parameters

// login
var activeUser = loadedUser = await user.login(
                    usernameVar: 'email',
                    usernameVal: 'test@test.com',
                    password: '12345678',
                  );

// check
var checkedUser = await user.check(
                    type: 'Bearer ',
                    token: activeUser['api_token'],
                  ); 

// logout
var logout = await user.logout(
                    type: 'Bearer ',
                    token: activeUser['api_token'],
                  );

// register
Map form = {
            "email" : "test@test.com",
            "phone" : "00000000",
            "name" : "mohamed sayed",
            "password" : "1234567",
            };
registeredUser = await user.register(form: form);
```


## Project Created & Maintained By

### [![Mohamed Sayed](./logo.png)](https://msayed.net)

Software Engineer | In :heart: with Flutter

# Donate

> If you found this project helpful or you learned something from the source code and want to thank me, consider buying me a cup of :coffee:
>
> - [PayPal](https://www.paypal.me/msayednet/)

## Note : All Contibutions Are Welcomed

CHANGELOG.md
## 1.2.2

* Change Package Template

## 1.2.1

* Easy change routes

## 1.2.0

* Enhanced Quality
* Enhanced Docs

## 1.1.1+5

* Enhanced Documentation

## 1.1.1

* new methods : `register`
* performance optimization

## Next

* Adding more feautres


lib/user_auth.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class UserAuth {
  // ? vars ? //
  String baseUrl;
  var _prefs;
  String loginUrl = "/user/login";
  String registerUrl = "/user/register";
  String checkUrl = "/user/details";
  String logoutUrl = "/user/logout";

  ///------------------------------------
  /// User : init
  ///------------------------------------
  /// initialize variables
  init({
    String apiBaseUrl,
    bool store = false,
    prefs,
    String loginUrl = "/user/login",
    String registerUrl = "/user/register",
    String checkUrl = "/user/details",
    String logoutUrl = "/user/logout",
  }) async {
    baseUrl = apiBaseUrl;
    if (store && prefs != null) {
      _prefs = prefs;
    }
    loginUrl = loginUrl;
    registerUrl = registerUrl;
    checkUrl = checkUrl;
    logoutUrl = logoutUrl;
  }

  ///------------------------------------
  /// User : Login
  ///------------------------------------
  /// return user data
  Future<dynamic> login({
    String usernameVar,
    String usernameVal,
    String password,
  }) async {
    try {
      // ---- API Call ---- //
      var response = await http.post(
        baseUrl + loginUrl,
        body: {
          usernameVar: usernameVal,
          'password': password,
        },
      );
      // ---- Response ---- //
      var data = json.decode(response.body);
      if (_prefs != null) {
        _prefs.setString('user', json.encode(data['user']));
      }
      return data['user'];
    } catch (e) {
      print('Error: ' + e.toString());
    }
  }

  ///------------------------------------
  /// User : Register
  ///------------------------------------
  /// return user data
  Future<dynamic> register({Map form}) async {
    try {
      // ---- API Call ---- //
      var response = await http.post(
        baseUrl + registerUrl,
        body: form,
      );
      // ---- Response ---- //
      var data = json.decode(response.body);
      if (_prefs != null) {
        _prefs.setString('user', json.encode(data['user']));
      }
      return data['user'];
    } catch (e) {
      print('Error: ' + e.toString());
    }
  }

  ///------------------------------------
  /// User : Load
  ///------------------------------------
  /// return user data
  Future<dynamic> loadUser() async {
    if (_prefs.getString('user') != null) {
      return json.decode(_prefs.getString('user'));
    } else {
      return null;
    }
  }

  ///------------------------------------
  /// User : Check
  ///------------------------------------
  /// return user data
  Future<dynamic> check({
    String type,
    String token,
  }) async {
    try {
      // ---- API Call ---- //
      var response = await http.post(baseUrl + checkUrl, headers: {
        'Authorization': type + token,
      });

      // ---- Response ---- //
      var data = json.decode(response.body);
      return data['user'];
    } catch (e) {
      print('Error: ' + e.toString());
    }
  }

  ///------------------------------------
  /// User : Logout
  ///------------------------------------
  Future<dynamic> logout({
    String type,
    String token,
  }) async {
    try {
      // ---- API Call ---- //
      var response = await http.post(baseUrl + logoutUrl, headers: {
        'Authorization': type + token,
      });

      // ---- Response ---- //
      if (response.statusCode == 200) {
        _prefs.clear();
        return true;
      } else {
        return false;
      }
    } catch (e) {
      print('Error: ' + e.toString());
    }
  }

  String placeHolder() {
    return 'Not Loaded';
  }
}


android/src/main/java/net/msayed/user_auth/UserAuthPlugin.java
package net.msayed.user_auth;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** UserAuthPlugin */
public class UserAuthPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "user_auth");
    channel.setMethodCallHandler(new UserAuthPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:async';
import 'package:user_auth/user_auth.dart';

// Note : change 'Bearer ' and api_token with your own parameters
// ---- user instance ----- //
UserAuth user = new UserAuth();

// ---- vars ---- //
var activeUser, checkedUser, loadedUser, registeredUser, loggedOut = false;

Future main() async {
  WidgetsFlutterBinding.ensureInitialized();

  // ---- user init ---- //
  SharedPreferences prefs = await SharedPreferences.getInstance();
  await user.init(
    apiBaseUrl: 'https://example.com/api',
    store: true,
    prefs: prefs, // need if store: true
    loginUrl: "/user/login",
    registerUrl: "/user/register",
    checkUrl: "/user/details",
    logoutUrl: "/user/logout",
  );

  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
    loggedOut = false;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin Example'),
        ),
        body: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              RaisedButton(
                child: Text('Load User'),
                onPressed: () async {
                  loadedUser = await user.loadUser();
                  loggedOut = false;
                  setState(() {});
                },
              ),
              Text('loaded..'),
              loadedUser != null
                  ? Text(loadedUser.toString())
                  : Text(user.placeHolder()),
              Divider(height: 40),
              RaisedButton(
                child: Text('Login'),
                onPressed: () async {
                  activeUser = loadedUser = await user.login(
                    usernameVar: 'email',
                    usernameVal: 'vendor@test.com',
                    password: '123456',
                  );
                  loggedOut = false;
                  setState(() {});
                },
              ),
              Text('login..'),
              activeUser != null
                  ? Text(activeUser.toString())
                  : Text(user.placeHolder()),
              Divider(height: 40),
              RaisedButton(
                child: Text('Check'),
                onPressed: () async {
                  checkedUser = await user.check(
                    type: 'Bearer ',
                    token: activeUser['api_token'],
                  );
                  setState(() {});
                },
              ),
              Text('check..'),
              checkedUser != null
                  ? Text(checkedUser.toString())
                  : Text(user.placeHolder()),
              Divider(height: 40),
              RaisedButton(
                child: Text('Register'),
                onPressed: () async {
                  Map form = {
                    "email": "test@test.com",
                    "phone": "00000000",
                    "name": "mohamed sayed",
                    "password": "1234567",
                  };
                  registeredUser = await user.register(form: form);
                  setState(() {});
                },
              ),
              Text('register..'),
              registeredUser != null
                  ? Text(registeredUser.toString())
                  : Text(user.placeHolder()),
              Divider(height: 40),
              RaisedButton(
                child: Text('Logout'),
                onPressed: () async {
                  loggedOut = await user.logout(
                    type: 'Bearer ',
                    token: activeUser['api_token'],
                  );
                  activeUser = checkedUser = loadedUser = null;
                  setState(() {});
                },
              ),
              Text('logout..'),
              Text(loggedOut.toString()),
            ],
          ),
        ),
      ),
    );
  }
}
